### PR TITLE
Optimize string handling in path search

### PR DESCRIPTION
**Pull request by osjc**
Use "+=" to construct the various game paths. Use a temporary
to extract and reuse a result of a few common subexpressions
to further reduce the overhead.

### CODE REVIEW

**Comment by osjc** [comment]
The new code is pretty assemblyish but that is what happens when you try to optimize a bunch of expressions that mix char[some-random-constant] with std:string.

File: src/content/subgames.cpp [hunk 1]
```diff
@@ -67,15 +67,22 @@ SubgameSpec findSubgame(const std::string &id)
 	std::vector<GameFindPath> find_paths;
 	while (!search_paths.at_end()) {
 		std::string path = search_paths.next(PATH_DELIM);
-		find_paths.emplace_back(path + DIR_DELIM + id, false);
-		find_paths.emplace_back(path + DIR_DELIM + id + "_game", false);
+		path += DIR_DELIM;
```
Comment (nerzhul) [inline src/content/subgames.cpp]: I prefer path.append(DIR_DELIM).append(id) to prevent multiple useless lines

File: src/content/subgames.cpp [hunk 2]
```diff
@@ -67,15 +67,22 @@ SubgameSpec findSubgame(const std::string &id)
...
 	}
-	find_paths.emplace_back(
-			user + DIR_DELIM + "games" + DIR_DELIM + id + "_game", true);
-	find_paths.emplace_back(user + DIR_DELIM + "games" + DIR_DELIM + id, true);
-	find_paths.emplace_back(
-			share + DIR_DELIM + "games" + DIR_DELIM + id + "_game", false);
-	find_paths.emplace_back(share + DIR_DELIM + "games" + DIR_DELIM + id, false);
+
+	std::string game_base = DIR_DELIM;
+	game_base += "games";
```
Comment (nerzhul) [inline src/content/subgames.cpp]: same here

File: src/content/subgames.cpp (hunk 1 above)
Comment (osjc) [reply src/content/subgames.cpp]: Tested, works, so here you are.

**Comment by osjc** [comment]
A less assemblyish version suggested by nerzhul. It works too so let it be.

File: src/content/subgames.cpp [hunk 3]
```diff
@@ -67,15 +67,19 @@ SubgameSpec findSubgame(const std::string &id)
 	std::vector<GameFindPath> find_paths;
 	while (!search_paths.at_end()) {
 		std::string path = search_paths.next(PATH_DELIM);
-		find_paths.emplace_back(path + DIR_DELIM + id, false);
-		find_paths.emplace_back(path + DIR_DELIM + id + "_game", false);
+		path = path.append(DIR_DELIM).append(id);
+		find_paths.emplace_back(path, false);
+		path = path.append("_game");
```
Comment (nerzhul) [inline src/content/subgames.cpp]: you don't need to affect the reference. Append modify the path object itself

http://www.cplusplus.com/reference/string/string/append/

Comment (osjc) [reply src/content/subgames.cpp]: At least SOMETHING behaves normally in C++. As a matter of fact, I disliked the "path = path.blabla" part of the new lines. Good to see that I can get rid of them.

Comment (osjc) [reply src/content/subgames.cpp]: Ok. Got rid of the pesky "path = path. ...".

**Pull request merged**